src/util/src/algebra.rs

```diff
@@ -0,0 +1,582 @@
+    pub fn pow<T>(&self, exponent: T, group: &Group) -> GroupElement
+    where
+        BigUint: From<T>,
+    {
+        let x = BigUint::from(exponent);
+        GroupElement(self.0.modpow(&x, &group.p))
+    }
+
+    /// Performs modular exponentiation of the group element with a given field element.
+    ///
+    /// This defines an action of the field over the group.
+    pub fn exp(&self, exponent: &FieldElement, group: &Group) -> GroupElement {
+        GroupElement(self.0.modpow(&exponent.0, &group.p))
+    }
+
+    /// Checks if the element is a valid member of the given group.
+    ///
+    /// This method return true iff `0 <= self < p` and `self^q % p == 1` where `p` is the group modulus and `q` the group order.
+    pub fn is_valid(&self, group: &Group) -> bool {
+        // It is enough to check the upper bound as self.0 is unsigned.
+        let elem_less_than_p = self.0 < group.p;
+        let elem_has_order_q = self.0.modpow(&group.q, &group.p).is_one();
+        elem_less_than_p && elem_has_order_q
+    }
+
+    /// Returns the left padded big-endian encoding of the group element.
+    ///
+    /// The encoding follows Section 5.1.1 in the specs.
+    pub fn to_be_bytes_left_pad(&self, group: &Group) -> Vec<u8> {
+        to_be_bytes_left_pad(&self.0, group.l_p())
+    }
+
+    pub fn to_biguint(&self) -> BigUint {
+        self.0.clone()
+    }
+}
+
+impl Group {
+    /// Constructs a new multiplicative integer group `Z_p^r`.
+    ///
+    /// The arguments are
+    /// - `modulus` - the modulus `p`
+    /// - `cofactor`- the cofactor `r`
+    /// - `generator` - a generator `g`
+    ///
+    /// This function checks that the group is valid according to [`Group::is_valid`].
+    /// Checking the validity of inputs is expensive.
+    /// A group should therefore be constructed once and then reused as much as possible.
+    ///
+    /// Alternatively, one can use fixed, *trusted/tested* parameters with [`Group::new_unchecked`].
+    pub fn new(
+        modulus: BigUint,
+        cofactor: BigUint,
+        generator: BigUint,
+        csprng: &mut Csprng,
+    ) -> Option<Self> {
+        let order = Self::compute_order(&modulus, &cofactor);
+        let group = Group {
+            p: modulus,
+            r: cofactor,
+            g: generator,
+            q: order,
+        };
+        if group.is_valid(csprng) {
+            return Some(group);
+        }
+        None
+    }
+
+    /// Constructs a new group without checking the validity according to [`Group::is_valid`].
+    pub fn new_unchecked(modulus: BigUint, cofactor: BigUint, generator: BigUint) -> Self {
+        let order = Self::compute_order(&modulus, &cofactor);
+        Group {
+            p: modulus,
+            r: cofactor,
+            g: generator,
+            q: order,
+        }
+    }
+
+    fn compute_order(modulus: &BigUint, cofactor: &BigUint) -> BigUint {
```

Comment:
Do we really want to calculate `q` from `r` like this? 
Maybe it would be easier to check the right relation between `p` and `q` in `is_valid()`?

Reply:
See the discussion in the above.